docs/architecture/adr-017-subspace-fee-grant.md

```diff
@@ -0,0 +1,320 @@
+1. the fees provider creates a group;
+2. the fees provider grants an allowance to the group;
+3. the user asks for a subspace group fee grant;
+4. the fee provider adds the user to the group;
+5. the user can use the service without any tokens inside the specified subspace, the fees will be paid by the fees provider.
+```mermaid
+sequenceDiagram
+    participant User
+    participant Fees Provider
+    participant Desmos
+    
+    Fees Provider->>Desmos: 1. create a group
+    Desmos-->>Fees Provider: return ok
+    Fees Provider->>Desmos: 2. grant an allowance to the group
+    Desmos-->>Fees Provider: return ok
+    
+    User->>Fees Provider: 3. ask for a subspace fee grant
+    Fees Provider->>Desmos: 4. add the user to the group
+    Desmos-->>Fees Provider:  return ok
+    Fees Provider-->>User: return ok
+
+    User->>Desmos: 5. use app without tokens
+    Desmos-->>User: return ok
+```
+
+### DeductFeeDecorator
+
+Currently, `x/auth` provides a `DeductFeeDecorator` based on `x/feegrant` to execute the action deducting the fees of a transaction from the signer/feepayer. We will build a new subspace-specified `DeductFeeDecorator` to replace the current one.
+
+The new subspace-specified `DeductFeeDecorator` will deal with the fees with the process as follows:
+1. check all the messages in the transaction are the subspace messages and all of them are to the same subspace;
+2. apply `x/subspaces` `DeductFeeDecorator` if the transaction contains subspace messages from the same subspace and the user/group fee grant exists, or apply `x/auth` `DeductFeeDecorator`;
+3. deduct the fees from the fee payer.
+
+```mermaid
+graph TD
+  id1([Start]) --> id2[Check transaction]
+  id2 --> id3{Are all subspace msgs<br/ >to the same subspace?}
+  id3 -- YES --> id4{Is the user/group grant exist?}
+  id3 -- NO --> id5[Apply `x/auth`<br />DeductFeeDecorator]
+  id4 -- YES --> id6[Apply `x/subspaces`<br />DeductFeeDecorator]
+  id4 -- NO --> id5
+  id5 --> id7[Deduct fees]
+  id6 --> id7
+  id7 --> id8[End]
+```
+
+### Types
+
+#### Allowance
+
+Currently, `x/feegrant` provides an interface `FeeAllowanceI` and the implementations for the interface like `BasicAllowance`, `AllowedMsgAllowance` and `PeriodicAllowance`. Each of them already have the operation process to deal with the allowance. We will reuse them to be the allowance object inside the `x/subspace`.
+
+#### Grant
+
+As `Allowance`, `x/feegrant` has `Grant` object to record all full context, which contains `granter`, `grantee` and what kind of `allowance` is granted. We will reuse it to store in the subspace namespace KVStore.
+
+#### GroupGrant
+
+To show the group grant, we will implement the `GroupGrant` object to record all full context containing `granter`, `group_id` and what kind of `allowance` is granted, the structure will be as follows:
```

Comment:
```suggestion
#### Group Grant

In order to represent a fee grant granted to a user group, we will implement a `GroupGrant` type. This will contain the `granter`, `group_id` and what kind of `allowance` is granted to the group:

```